### PR TITLE
Support provider-side preview.

### PR DESCRIPTION
**Pull request by pgavlin**
Read the proposed state out of the diff.

### CODE REVIEW

**Comment by pgavlin** [comment]
These changes also include a local implementation of the field reader used to read values out of the Terraform diff.

The field reader implementations in the plugin SDK do not properly handle nested unknown values, instead treating such values as empty. The local implementation of the field reader correctly leaves nested unknown values as unknown.

A local implementation was chosen instead of an implementation in the plugin SDK fork so that it could be easily scoped to reading values from an `*InstanceDiff`, which only happens during previews. Naively changing the upstream field reader would affect all field reads, including those that happen during updates. This seems unacceptably risky. As an alternative, we could add a new implementation of the field reader in our fork, which may simplify any licensing issues.

**Comment by pgavlin** [comment]
I have verified that this fixes https://github.com/pulumi/pulumi-azure/issues/192.

File: pkg/tfbridge/provider.go [hunk 1]
```diff
@@ -954,23 +964,30 @@ func (p *Provider) Update(ctx context.Context, req *pulumirpc.UpdateRequest) (*p
...
+		}
+		if newstate.ID() == "" {
+			return nil, fmt.Errorf("expected non-empty ID for new state during Update of %s", urn)
+		}
 		if err != nil {
-			return nil, err
+			reasons = append(reasons, errors.Wrapf(err, "updating %s", urn).Error())
+		}
+	} else {
+		newstate, err = diff.ProposedState(res.TF, nil)
```
Comment (leezen) [inline pkg/tfbridge/provider.go:987]: I haven't spent a lot of time in this code, but given the way this is structured, I would've expected to pass in the resource's prior state here as opposed to `nil` for previewing an update?

**Review by leezen — COMMENTED** [review]
Any chance you had an opportunity to validate that this fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/251?

File: pkg/tfbridge/provider.go (hunk 1 above)
Comment (pgavlin) [reply pkg/tfbridge/provider.go:987]: Indeed, the prior state should be passed in. Still working on tests for this PR, clearly :)

Comment (leezen) [reply pkg/tfbridge/provider.go:987]: Sure thing. Ping me when you're ready for another pass.

**Pull request merged**